lib/services/expiry/expiry.go

```diff
@@ -0,0 +1,250 @@
+	"github.com/gravitational/teleport/lib/events"
+	"github.com/gravitational/teleport/lib/services"
+	"github.com/gravitational/teleport/lib/utils/interval"
+	"github.com/gravitational/trace"
+	"github.com/jonboulle/clockwork"
+)
+
+var (
+	// scanInterval is the interval at which the expiry checker scans for access requests
+	scanInterval              = time.Minute * 5
+	pendingRequestGracePeriod = time.Second * 40
+)
+
+const (
+	semaphoreName       = "expiry"
+	semaphoreExpiration = time.Minute
+	semaphoreJitter     = time.Minute
+
+	// minPageDelay is the minimum delay between processing each page of access requests
+	minPageDelay           = time.Millisecond * 200
+	accessRequestPageLimit = 100
+	maxExpiresPerCycle     = 120
+)
+
+// Config provides configuration for the expiry server.
+type Config struct {
+	// Log is the logger.
+	Log *slog.Logger
+	// Emitter is an events emitter, used to submit discrete events.
+	Emitter apievents.Emitter
+	// AccessPoint is a expiry access point.
+	AccessPoint authclient.ExpiryAccessPoint
+	// Clock is the clock used to calculate expiry.
+	Clock clockwork.Clock
+	// HostID is a unique ID of this host.
+	HostID string
+}
+
+// CheckAndSetDefaults checks required fields and sets default values.
+func (c *Config) CheckAndSetDefaults() error {
+	if c.Emitter == nil {
+		return trace.BadParameter("no Emitter configured for expiry")
+	}
+	if c.AccessPoint == nil {
+		return trace.BadParameter("no AccessPoint configured for expiry")
+	}
+	if c.Clock == nil {
+		return trace.BadParameter("no Clock configured for expiry")
+	}
+	return nil
+}
+
+// Service is a expiry service.
+type Service struct {
+	*Config
+
+	ctx      context.Context
+	cancelfn context.CancelFunc
+}
+
+// New initializes a expiry service
+func New(ctx context.Context, cfg *Config) (*Service, error) {
+	if err := cfg.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	s := &Service{
+		Config: cfg,
+		ctx:    ctx,
+	}
+	return s, nil
+}
+
+// Run starts the expiry service.
+func (s *Service) Run() error {
+	semCfg := services.SemaphoreLockConfigWithRetry{
+		SemaphoreLockConfig: services.SemaphoreLockConfig{
+			Service: s.AccessPoint,
+			Params: types.AcquireSemaphoreRequest{
+				SemaphoreKind: types.KindAccessRequest,
+				SemaphoreName: semaphoreName,
+				MaxLeases:     1,
+				Holder:        s.HostID,
+			},
+			Expiry: semaphoreExpiration,
+			Clock:  s.Clock,
+		},
+		Retry: retryutils.LinearConfig{
+			Clock:  s.Clock,
+			First:  time.Second,
+			Step:   semaphoreExpiration / 2,
+			Max:    semaphoreExpiration,
+			Jitter: retryutils.DefaultJitter,
+		},
+	}
+
+	poll := interval.New(interval.Config{
+		Duration:      scanInterval,
+		FirstDuration: retryutils.FullJitter(scanInterval),
+		Jitter:        retryutils.SeventhJitter,
+		Clock:         s.Clock,
+	})
+	defer poll.Stop()
+
+	for {
+		lease, err := services.AcquireSemaphoreLockWithRetry(
+			s.ctx,
+			semCfg,
+		)
```

Comment:
Nit: Saved you 3 LoC.
```suggestion
		lease, err := services.AcquireSemaphoreLockWithRetry(s.ctx, semCfg)
```